[JENKINS-66771] Add a "conflictsWith" concept to agent RetentionStrategy…

jimklimov opened the pull request:
…(goes with inDemandDelay handling)

See [JENKINS-66771](https://issues.jenkins-ci.org/browse/JENKINS-66771) for detailed rationale.

The short take-away is: this PR adds an ability to run many agents (e.g. chroots or containers with various build/test environments) on a weaker host that can not execute all those environment at the same time - but can process them one at a time.

Practical use-case - many Linux containers with QEMU architectures, slow and resource consuming at run-time but can easily be pre-defined and stored on the same cloud host dedicated for a project.

Note that I sadly did not make any test-cases, just not proficient enough with Java to make some from scratch in finite time. Usually I would extend an existing test, but seems there is none for `inDemandDelay` and `idleDelay` either. So would have to defer this to anyone who can.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Add a "conflictsWith" concept to agents starting on demand

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://www.jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

jimklimov committed:
[JENKINS-66771] Add a "conflictsWith" concept to RetentionStrategy.java (goes with inDemandDelay handling)

jimklimov committed:
Merge branch 'master' into JENKINS-66771

jimklimov committed:
Fix input field type for retentionStrategy.conflictsWith

timja commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -261,9 +307,14 @@ public long check(final SlaveComputer c) {
                 if (needComputer) {
                     // we've been in demand for long enough
-                    logger.log(Level.INFO, "Launching computer {0} as it has been in demand for {1}",
+                    if (hasConflict.size() > 0) {
```
```suggestion
                    if (!hasConflict.isEmpty()) {
```

jimklimov replied:
fixed

jimklimov committed:
RetentionStrategy.java: for systems that enabled a conflictsWithPattern, log that there was no blocker when starting on demand

timja commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -217,13 +238,38 @@ public long getIdleDelay() {
+         *
+         * @return Value for property 'conflictsWith'.
+         */
+        public String getConflictsWith() {
```
add a `doCheckConflictsWith` method to validate the regex?

jimklimov replied:
fixed

timja commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -193,10 +199,25 @@ public String getDisplayName() {
+
         @DataBoundConstructor
-        public Demand(long inDemandDelay, long idleDelay) {
+        public Demand(long inDemandDelay, long idleDelay, String conflictsWith) {
```
generally optional fields should be created by a setter (and annotated by `@DataboundSetter`

jimklimov replied:
fixed

jimklimov committed:
RetentionStrategy.java: optimize condition check for conflictsWithPattern handling

Thanks @timja!

jimklimov committed:
RetentionStrategy.java: add doCheckConflictsWith() method

...and double-check at run-time (bad data loaded from config.xml etc.)
that the regex is valid before we use it

jimklimov committed:
Fix input field check for retentionStrategy.conflictsWith

jimklimov committed:
conflictsWith.html: document a "negative look-ahead" regex to simplify copy-pasting

jimklimov committed:
RetentionStrategy.java: try to log not-starting due to conflict into agent log visible from UI

jimklimov committed:
RetentionStrategy/Demand/config.jelly: define field conflictsWith a bit differently

Follow https://www.jenkins.io/doc/developer/forms/form-validation/

jimklimov committed:
RetentionStrategy.java: assign conflictsWith by @DataBoundSetter, not constructor

jimklimov commented:
Fixed the review comments (thanks @timja!) and some locally found imperfections.

Did not address auto-testing :\

jimklimov committed:
RetentionStrategy.java: report other Launching (inDemand) and Disconnecting (inIdle) cases into connection log of the agent

jimklimov committed:
RetentionStrategy.java: report computer name bracketed (when Launching and Disconnecting) to match possible list from conflictsWith regex hit

jimklimov committed:
RetentionStrategy.java: report Launching into connection log of the agent AFTER requesting the launch (which flushes the log)

daniel-beck reviewed (commented):
Seems like this might be better off in a plugin implementing a more powerful `RetentionStrategy` than what is in core? The use case looks like it would be addressed well by this change, but seems a bit narrow for supporting it in core.

jimklimov committed:
RetentionStrategy.java: if "conflictsWith" was not good, make sure "conflictsWithPattern" is neuteralized

jimklimov committed:
RetentionStrategy.java: evaluate "conflictsWithPattern" always for an online/onlining "other" computer; only constrain to computers available for processing for counting executors

jimklimov committed:
Merge remote-tracking branch 'upstream/master' as of 2021-10-22 into JENKINS-66771

jimklimov commented:
Gentle bump :)

This new strategy proved invaluable at the NUT CI farm, and needed no changes over the past month, so I'd love to not have to roll out forked builds but to use the vanilla upstream with this feature ;)

daniel-beck commented:
Does https://github.com/daniel-beck/conflict-aware-ondemand-retention-strategy-plugin help?

jimklimov commented:
Didn't try it yet, but logically looks cool! :)

If installed as plugin, would the tweak be configurable for the different
agent types (ssh, swarm, ...) as the current PR is?

Jim

On Fri, Dec 3, 2021, 11:51 Daniel Beck ***@***.***> wrote:

> Does
> https://github.com/daniel-beck/conflict-aware-ondemand-retention-strategy-plugin
> help?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/5764#issuecomment-985419500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMPTFDNQZQ2NYRYD7XC23DUPCOM7ANCNFSM5FDVVUGA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

daniel-beck commented:
> If installed as plugin, would the tweak be configurable for the different agent types (ssh, swarm, ...) as the current PR is?

I expect so, agent launcher and retention strategy should be independent. As I wrote, I haven't tested in beyond making sure the UI works.

jimklimov commented:
Sorry for prolonged delay with the end of year rush, but I finally got a moment to set that system up with your plugin version of my PR and a "vanilla" LTS Jenkins core (2.319.1).

Took some time to find the regexes to apply from backup of older configs and click up the configuration to use the new ondemand method in a couple dozen agents ;)

Finally queued some builds that should hit the QEMU VMs so will see by tomorrow how it goes.

jimklimov commented:
@daniel-beck : as for repo "transfer", maybe better it could make way into jenkins-ci org and add me as maintainer to brush up the README, publish releases, etc?

I'll wait for the builds noted above to progress further (e.g. that it would hand over to next agents when initially started ones complete), but so far it looks promising that the feature can live outside the Jenkins core - thanks for the help!

daniel-beck commented:
@jimklimov 

> as for repo "transfer", maybe better it could make way into jenkins-ci org and add me as maintainer to brush up the README, publish releases, etc?

Delete your fork and I'll transfer the repo to you, then you can request hosting. (Hosting works best when there are no forks to begin with.)

daniel-beck commented:
Closing this PR for now given where this seems to be going.

jimklimov commented:
> Delete your fork and I'll transfer the repo to you, then you can request hosting. (Hosting works best when there are no forks to begin with.)

Thanks, nuked :)
